### PR TITLE
iOS: Prevent accidental block hover

### LINKED ISSUE #2601: Enter in iOS Safari hides keyboard
This breaks the writing flow. I have to re-focus the area to be able to type further.

### PR DESCRIPTION
**Pull request by tg-ephox**
## Description

Fixes https://github.com/WordPress/gutenberg/issues/2601

- Adds iOS momentum scrolling, where previously scrolling would abruptly stop once the finger was lifted from the screen.
- Fixes the following reported behavior:
> Also have to click twice in a block to be able to get a caret and keyboard. May be related.

## How Has This Been Tested?

Manual testing on device

## Types of changes

Bug fix

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style.
- [x] My code follows has proper inline documentation.

### CODE REVIEW

**Comment by ephox-mogran** [comment]
With respect to the keyboard disappearing on enter, we've found a few things that might be helpful:

a) it happens when a block node is being split. You can replicate it easily on a new post by pressing enter halfway through a paragraph block.
b) the keyboard is most likely disappearing due to a blur happening. IOS is quite opinionated about when the keyboard should be visible ... a mere focus call isn't always enough to bring it back. It also tends to avoid showing a keyboard when it feels the focus doesn't require it. There are two sections of the code that if you remove their focus logic, the keyboard does not go away.

in `visual-editor/block.js` in `componentDidMount`. This focus call is firing a focus on a div that while it may be focusable, is not a container that allows text entry, therefore IOS probably feels it should hide the keyboard.

in `blocks/editable/index.js` in `updateFocus`. There is a blur call here. A blur call instantly tells IOS to hide the keyboard (in fact, it's often what developers use to intentionally hide the keyboard). Sometimes, a subsequent focus call isn't enough to bring back the keyboard. It tends to (sometimes) depend on:

* whether the focus call is within the same event queue (sort of like what is mentioned in https://developer.mozilla.org/en-US/docs/Web/JavaScript/EventLoop) 
* where the focus was previously
* whether the newly focused element accepts keyboard input
* how close you are to your project deadline

As you can probably infer, a lot of this is empirically determined, rather than read from a spec somewhere. If you know where the spec is, though, please tell us :) 

So in summary, if you remove both the focus and the blur line, the keyboard stays open and the block is correctly focused. I'm not suggesting that's the solution at all ... just providing some information that might help explain what's going on. I tried to replicate the issue in a fiddle outside of Gutenberg, but had no luck. It probably depends on when React is firing focus and blur on these nodes. The sequence seems to be:  `previous.blur > block.focus > editable.focus`, but replicating that order wasn't sufficient to recreate it elsewhere.

File: editor/layout/style.scss [hunk 1]
```diff
@@ -22,6 +22,11 @@
 	overflow-y: auto;
 }
 
+.mobile .editor-layout__editor {
+	overflow-y: scroll;
+	-webkit-overflow-scrolling: touch;
+}
```
Comment (mcsf) [inline editor/layout/style.scss]: [MDN seems very wary](https://developer.mozilla.org/en-US/docs/Web/CSS/-webkit-overflow-scrolling) of using `-webkit-overflow-scrolling`. Are we safe? Is there a risk of targeting unwanted platforms where this could have a bad impact? We could target this to iOS.

**Comment by mcsf** [comment]
Thanks for the comprehensive comment, @ephox-mogran.

@tg-ephox, could you provide full testing instructions, as well as specifics on what this should be fixing? I'm not able to repro some of the problems mentioned in the parent issue, at least not on an iPhone—can grab an iPad for a fuller review. Thanks.

File: editor/layout/style.scss (hunk 1 above)
Comment (tg-ephox) [reply editor/layout/style.scss]: @mcsf only webkit should interpret this from the vendor prefix and body will have the `.mobile` class on for mobile devices? Have tested on iPad and Nexus tablet.

**Comment by tg-ephox** [comment]
@mcsf the fix is for 'Also have to click twice in a block to be able to get a caret and keyboard. May be related.' that @iseulde reported in #2601
I added the iOS momentum scrolling as it was quite jarring to use as an iOS user!

File: editor/layout/style.scss (hunk 1 above)
Comment (mcsf) [reply editor/layout/style.scss]: > only webkit should interpret this

Indeed, but my concern was whether this could affect non-iOS WebKit agents negatively. However, if you've tested on a Nexus and feel confident about this, I can withdraw my question.

**Review by mcsf — APPROVED** [review]
This is good to go as a smaller PR. I notice the _In Progress_ label is present, but we can defer any other work to a new PR.

**Comment by iseulde** [comment]
For the scrolling, there was also #2860.

File: editor/layout/style.scss (hunk 1 above)
Comment (iseulde) [reply editor/layout/style.scss]: Should we really target `.mobile` here? What is the connection between the two? Can we not just add it to `.editor-layout__editor`?

**Comment by mcsf** [comment]
Since #2860 is slightly broader in reach and has been previously tested and approved, let's rebase and merge that one, and we can amend this one to only deal with the hover issue.

**Comment by tg-ephox** [comment]
No problems @mcsf . Sorry @iseulde I didn't see #2860 and just made a fix while I was on the iOS stuff.

File: editor/modes/visual-editor/block.js [hunk 2]
```diff
@@ -72,6 +72,8 @@ class VisualEditorBlock extends Component {
 		this.onKeyDown = this.onKeyDown.bind( this );
 		this.onBlockError = this.onBlockError.bind( this );
 		this.insertBlocksAfter = this.insertBlocksAfter.bind( this );
+		this.onTouchStart = this.onTouchStart.bind( this );
+		this.onClick = this.onClick.bind( this );
```
Comment (iseulde) [inline editor/modes/visual-editor/block.js]: What would be the diff between touch start and click here? Might it be better stick to touch events entirely to set `this.hadTouchStart`?

**Comment by iseulde** [comment]
It would be nice to have some inline comments for this browser fix. I will test in just a bit on a device.

**Comment by iseulde** [comment]
This works for me on iOS. I'm surprised mouse enter/move fire there. :/

File: editor/modes/visual-editor/block.js (hunk 2 above)
Comment (tg-ephox) [reply editor/modes/visual-editor/block.js]: I tried onTouchEnd but it fires too soon it needs to be after mousemove

**Comment by iseulde** [comment]
This PR closed #2601 while only one non related issue was fixed.
